### ISSUE TITLE
How to reuse connection when using DSL?

### ISSUE THREAD

**Issue by Jeff17Robbins** [graphql-python/gql#138]
Each DSL operation calls `self.client.execute` which apparently creates a new connection each time.  Passing a `session` to `DSLSchema` doesn't work and causes an infinite recursion.

```
Traceback (most recent call last):
  File "gql_test2.py", line 45, in <module>
    query = ds.Query.listCasess().select(ds.ModelCasesConnection.items.select(ds.Cases.id))
  File "C:\Program Files\Python38\lib\site-packages\gql\dsl.py", line 38, in __getattr__
    type_def = self.schema.get_type(name)
  File "C:\Program Files\Python38\lib\site-packages\gql\dsl.py", line 38, in __getattr__
    type_def = self.schema.get_type(name)
  File "C:\Program Files\Python38\lib\site-packages\gql\dsl.py", line 38, in __getattr__
    type_def = self.schema.get_type(name)
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded
```

Is there some known good way to reuse the connection when using DSL?  Is the `session` meant to work, and this is simply a bug?

**Comment by leszekhanusz** [comment]
This is not possible to reuse the connection with the current DSL code.
The DSL code was made before all the 3.0 changes and is still quite experimental.
A refactor is probably needed so that it only returns a `DocumentNode` instead of needing to execute the query itself, this way we could execute the query how we want (in a session for example)

**Comment by Jeff17Robbins** [comment]
That makes sense.  For now, I'm doing this:

**Comment by Jeff17Robbins** [comment]
That makes sense.  For now, I'm doing this

```
    query = f'mutation {{ {str(_query)} }}' # where _query is built using the DSL
    res = session.execute(gql(query))
```

If I understood you properly, in the future, it might be possible to do this

```
    res = session.execute(_query)
```

and avoid the trip back to text and parsing?

**Comment by leszekhanusz** [comment]
> If I understood you properly, in the future, it might be possible to do this
> 
> ```
>     res = session.execute(_query)
> ```
> 
> and avoid the trip back to text and parsing?

Yes, exactly, something like this.

**Comment by leszekhanusz** [comment]
I refactored the DSL code and the proposed way is a little bit different.
Example:
```python
query = dsl_gql(ds.Query.hero.select(ds.Character.name))                                                            
result = session.execute(query)
```

**Comment by leszekhanusz** [comment]
Fixed in v3.0.0a5
